### PR TITLE
AP_Scripting: implement tricks on a switch on top of trajectory tracking

### PR DESCRIPTION
**Pull request by tridge**
this re-implements tricks on a switch with the new accurate trajectory tracking code.

It adds new parameters:

 - TRIK_ENABLE=0/1
 - TRIK_COUNT for number of tricks
 - TRIK_ACT_FN for the rc option to use to activate tricks (default 300)
 - TRIK_SEL_FN for the rc option to use to select which trick (default 301)

So if you want to use a 3 position switch on RC7 to activate and use a knob on RC8 for selection then you would set:

- RC7_OPTION = 300
- RC8_OPTION = 301

then if tricks are enabled the following parameters are created per trick:

 - TRIKn_ID
 - TRIKn_ARG1
 - TRIKn_ARG2
 - TRIKn_ARG3
 - TRIKn_ARG4

You can have a maximum of 11 tricks.

The ID numbers are from the trajectory table. The arguments are path specific.

### CODE REVIEW

File: libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua [hunk 1]
```diff
@@ -1285,6 +1338,147 @@ function create_auto_mission()
...
+--]]
+function get_trick_selection()
+   if trick_sel_chan == nil then
+      trick_sel_chan = rc:find_channel_for_option(TRIK_SEL_FN:get())
+      if trick_sel_chan == nil then
+         return 0
+      end
+   end
+   -- get trick selection based on selection channel input and number of tricks
+   local i = math.floor(TRIK_COUNT:get() * 0.5*(trick_sel_chan:norm_input_ignore_trim()+1))+1
```
Comment (Hwurzburg) [inline libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua:1408]: I think this needs to be changed...wont work with a normal three position switch or pot if the high position is even slightly less than RCx_MAX....bottom is okay...

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
very cool....cant wait to fly it

File: libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua (hunk 1 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua:1408]: found this...dont know if its in our LUA
```suggestion
   local i = math.floor(TRIK_COUNT:get() * 0.5*(trick_sel_chan:norm_input_ignore_trim()+1))+1
   i = math.clamp(i, 1, TRIK_COUNT:get())
```

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

File: libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua (hunk 1 above)
Comment (Hwurzburg) [reply libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua:1408]: also cruise alt hold is messed up I think...it creeps....

Comment (tridge) [reply libraries/AP_Scripting/examples/Aerobatics/Trajectory/plane_aerobatics.lua]: thanks, fixed and tested, going beyond min/max now works

**Comment by tridge** [comment]
@Hwurzburg I've fixed cruise abort to lock to the alt you abort at, which I think is the right behaviour

**Comment by Hwurzburg** [comment]
here is the bug that I spoke about....Trick6 was aborted, but TECS control was not returned
Screenshot from 2022-10-26 07-19-22
log:https://www.dropbox.com/s/y8b05n4vqse1v4l/aerobatics_fails.bin?dl=0

**Review by Hwurzburg — APPROVED** [review]

**Comment by Hwurzburg** [comment]
Trim aispeed is super critical for success...needs to be emphasized in setup...manuevers you could do manually at lower speeds dont work with this...once I increased to 16m/s things got a lot better....
observations: with these params loops and other manuevers exit a bit low and are not as tight in the paths...many at pretty good though...in the aero1 log, note the axial roll exit ...its really ugly and a bit out of control....also, as far as I can tell from the code and flying banked circle and climbing circle are exactly the same and one is redundant
log for first 10 tricks:https://www.dropbox.com/s/p5f2a8pzqh3xx8l/aero1.BIN?dl=0
log for remaining tricks:https://www.dropbox.com/s/7buy1f1bfg1xxvj/aero2.BIN?dl=0
then ran the test suite
the test suite aborts early....notice that mode takeoff navigation is wonky....that stabilization of axes has shudders....log: https://www.dropbox.com/s/mbgw2mr7nqyytpu/testaerosuite.BIN?dl=0

**Comment by tridge** [comment]
@Hwurzburg we've reproduced the roll on spiral climb, but haven't managed to fix it yet. Something odd with quaternion rotations, but don't know what it is yet

**Comment by tridge** [comment]
@Hwurzburg thanks to some help from @priseborough I have pushed a fix for the climbing circle
the banked circle has been removed

**Pull request merged**